### PR TITLE
Enable Lexilla's lexer for MS SQL Server 

### LINKED ISSUE #5940: SQL Language Colorizing Display Issue
### Description of the Issue

When using a single quote within an AS bracket, 

SELECT FIELD1 AS [FIELD'1]

The colorization is thrown off and turns grey. Some queries contain single quotes such as [Child's Name] for report display purposes in SSRS.

### Steps to Reproduce the Issue

1. Simply copy and paste this code into the editor and change the language indicator to SQL if needed and notice the color display. Once the single quote is removed, the color returns to normal for SQL syntax

SELECT FIELD1 AS [FIELD1]
            ,FIELD2 AS [FIELD2]
	    ,CHILD1_NAME AS [Child1 Name]
	    ,CHILD2_NAME AS [Child's Name]
	    ,FIELD3 AS [FIELD3]
FROM TABLE1
;

### Expected Behavior

Color should be normal SQL color when single quotes exist inside of AS brackets.
### Actual Behavior

All text after the single quote turned grey.

### Debug Information

SQLDisplayIssue

### LINKED ISSUE #7988: T-SQL missing keywords
### Description of the Issue
some T-SQL keywords and function are not highlithed:

APPROX_COUNT_DISTINCT
STRING_AGG
string_split
PARSE
TRY_CAST
TRY_CONVERT
TRY_PARSE
choose
iif
DATEDIFF_BIG 
DATEFROMPARTS 
DATETIME2FROMPARTS 
DATETIMEFROMPARTS 
DATETIMEOFFSETFROMPARTS 
EOMONTH 
SMALLDATETIMEFROMPARTS 
TIMEFROMPARTS 
openjson

### Steps to Reproduce the Issue

.1 open a new windows
.2 select SQL language
.3 Copy and past the previous lists in Notepad++

### Expected Behavior

When the selected language is SQL, the keywords should become highlited.

### Actual Behavior

When the selected language is sql, the text are not highlited.

### Debug Information

Notepad++ v7.8.4   (32-bit)
Build time : Jan 29 2020 - 01:33:47
Path : C:\Program Files (x86)\Notepad++\notepad++.exe
Admin mode : OFF
Local Conf mode : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1903
OS Build : 18362.657
Plugins : ComparePlugin.dll mimeTools.dll NppConverter.dll NppExport.dll Translate.dll XMLTools.dll

### LINKED ISSUE #11718: Nested block of comments in SQL code isn't highlighted correctly
### Description of the Issue
Code of SQL commented by block commands which is nested (example below) is not highlighted correctly. Please notice that in Microsoft SQ Server Management Studio below works as expected.

/* aaaaaaaaaa
   /*  bbbbbbbbbbbbbbb  */
aaaaaaaaaa */

### Steps to Reproduce the Issue

1.  Put below into editor
/* aaaaaaaaaa
   /*  bbbbbbbbbbbbbbb  */
aaaaaaaaaa */

2. Change highlighting with SQL
3. 

### Expected Behavior
all lines coming from example should be like as commented

### Actual Behavior
last line is incorrectly highlighted - like would be not commented

### LINKED ISSUE #12610: SQL Syntax Highlighter Issue with CASE statement
### Description of the Issue
SQL file syntax highlighter fails when a case is involved in a specific way

### Steps to Reproduce the Issue

1. Create new File
2. Paste the following SQL code:
      `UPDATE U SET
	      Title = xt.Title
      OUTPUT 
			      case when (inserted.nt_account is not null and charindex('\', inserted.nt_account) > 0) then
				      substring(inserted.nt_account, 1, charindex('\', inserted.nt_account) - 1)
			      else
				      null
			      end domain
      INTO dbo.TABLE
      FROM
	      dbo.TABLE2 xt`
4. Save as SQL file type

### Expected Behavior
Code after chardindex should not be gray as it's not comments

### Actual Behavior
You will notice the code after charindex is GRAY as if they were comments but they are not

image

### LINKED ISSUE #13160: SQL Language Option - Apostrophe ' in Square Brackets [] not escaped
### Description of the Issue
Apostrophes are not escaped in square brackets, resulting in the syntax highlighting for .sql files breaking.

### Steps to Reproduce the Issue
The below code adds a column with an apostrophe in the name.
--alter table Default_Test_Table add [bad_column'] [nvarchar](50) NULL
update Default_Test_Table set [bad_column'] = 'asdf'
update Default_Test_Table set [bad_column''] = 'asdf'

### Expected Behavior
In other editors like SSMS (tested with SQL Server Management Studio 19), this code will have the proper syntax highlighting, and the second line will execute correctly.

### Actual Behavior
Notepad++ doesn't escape the ', resulting in syntax highlighting breaking after that point in the file (in the above example, the third line will display incorrectly).

### Debug Information
Notepad++ v8.4.9   (64-bit)
Build time : Jan 27 2023 - 03:11:16
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : "C:\[REDACTED].sql"
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 21H2
OS Build : 19044.2604
Current ANSI codepage : 1252
Plugins : 
    mimeTools (2.9)
    NppConverter (4.5)
    NppExport (0.4)

First screenshot is Notepad++ (third line not displaying correctly)
NPP_01

Second is SSMS (displaying correctly)
NPP_02

### PR DESCRIPTION
**Pull request by rdipardo**
Notepad++ only officially supports Microsoft Windows, but does not have built-in support for Microsoft's SQL dialect, [Transact-SQL], used by Microsoft SQL Server.

This enables Lexilla's [dedicated MS SQL lexer].

To preserve the current file extension mapping, the new lexer is mapped to the `*.tsql` file extension only.

Users can select "Microsoft Transact-SQL" from the <kbd>Language</kbd> menu to apply MS SQL Server style to scripts with the `*.sql` extension.

- Fixes #5940

  issue5940-fixed

- Fixes #11718

  issue11718-fixed

- Fixes #12610 (see esp. https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12610#issuecomment-1340297100)

  issue12610-fixed

- Fixes #13160

  issue13160-fixed

- Closes #7988 (with 0bec5cb7)

[Transact-SQL]: https://learn.microsoft.com/en-us/sql/t-sql/language-reference
[dedicated MS SQL lexer]: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/lexilla/lexers/LexMSSQL.cxx

### CODE REVIEW

**Comment by ArkadiuszMichalski** [comment]
@rdipardo 

Any plan to ad autocomplete file? At least with the same words as in `langs.model.xml`?

**Comment by rdipardo** [comment]
> Any plan to add autocomplete file? At least with the same words as in `langs.model.xml`?

I think just implementing the lexer breaks enough ground without getting into that.

All the linked issues are just lexing issues, and I'm not sure I've completely figured out AppVeyor's commit filter syntax  &#x2014; i.e. why did [this job fail] and what is meant by "Changed files on this commit require full build"?

#11719 has more to do with auto-completion; somebody with better XML skills can link it to a future PR.

[this job fail]: https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/46290723

**Comment by ArkadiuszMichalski** [comment]
>All the linked issues are just lexing issues, and I'm not sure I've completely figured out AppVeyor's commit filter syntax — i.e. why did [this job fail](https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/46290723) and what is meant by "Changed files on this commit require full build"?

https://github.com/notepad-plus-plus/notepad-plus-plus/wiki/FAQ#how-to-create-a-non-blocking-pull-request-pr

When build we fetch full PR (with all changes made so far from all commits).
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/46290101/job/yuiwwsl2lvv96gs3
```
Build started
git clone -q https://github.com/notepad-plus-plus/notepad-plus-plus.git C:\projects\notepad-plus-plus
git fetch -q origin +refs/pull/13184/merge:
git checkout -qf FETCH_HEAD
```
If there have been any other commits up to this point that have changed files  that requiring a full build, then adding `[xml]` won't work because a full build is required. Unfortunately there is no way to determine changed files by a specific/last commit. We analyze the whole PR (it's something like a patch) and this includes all the changes.

Basically we don't have the entire branch with access to all its commits. If it can be somehow downloaded so that we have access to its commits to determine the changed files, it could be improved a bit. But that's a matter for a separate issue.

**Comment by dfs-** [comment]
@rdipardo Thanks for your effort. I tried to get it running using the AppVeyor Build, but failed. What I did was:
* placing the newly created .exe in the Notepad++ directory
* updated langs.xml and stylers.xml based on your changes in the according .model. file
* used a *.tsql file

I received no highlighting and can not choose the Transact-SQL language from the language menu. What am I doing wrong?

**Comment by rdipardo** [comment]
@dfs-,
> I received no highlighting and can not choose the Transact-SQL language from the language menu. What am I doing wrong?

Just run this Batch script; it will set everything up for you: [try_mssql.cmd.txt](https://github.com/notepad-plus-plus/notepad-plus-plus/files/10912946/try_mssql.cmd.txt)

**Comment by dfs-** [comment]
Thank you. This works great with the procedures I could test it with. I somewhat dislike some color choices, but that's up to my own stylers configuration so not an issue with the PR.

Would love this one added to N++ in the next release!

**Comment by rdipardo** [comment]
> I somewhat dislike some color choices

Those were copied directly from [a SciTE contributor's config file][0]. They're not pretty, but useful for demo purposes. Tweaking them in this PR means an hour-long build since [at least one commit][1] touches the C++ code.

If this PR is still sitting here by the time the artifacts expire, I might just push updated styles for the sake of fresh binaries.

[0]: https://www.scintilla.org/mssql.properties
[1]: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/13184#issuecomment-1439344141
